### PR TITLE
Major rewrite of promise support

### PR DESCRIPTION
**Pull request by nyurik**

### CODE REVIEW

**Comment by nyurik** [comment]
I think we should really switch to spaces after this - I cannot stand tabs in code, makes it much harder to edit

File: src/DataLoader.js [hunk 1]
```diff
@@ -29,52 +29,76 @@ module.exports = function ( createPromise, mwApi, clientStore, title, debounce,
...
+		// FIXME: we need to fix this horrid hack
+		// http://stackoverflow.com/questions/39970101/combine-multiple-debounce-promises-in-js
+		function setPromises( groupsToLoad, values, err ) {
+			var i, promise;
 
-				for ( i = 0; i < groupsToLoad.length; i++ ) {
-					loader.promiseByGroup[ groupsToLoad[ i ] ].resolve( rawMapData[ groupsToLoad[ i ] ] );
+			for ( i = 0; i < groupsToLoad.length; i++ ) {
+				promise = loader.promiseByGroup[ groupsToLoad[ i ] ];
+				if ( promise.mwResolve ) {
```
Comment (juliengirault) [inline src/DataLoader.js]: You're using `mwResolve` here but  `wmResolve` in `fetchGroup`

File: src/DataLoader.js [hunk 2]
```diff
@@ -29,52 +29,76 @@ module.exports = function ( createPromise, mwApi, clientStore, title, debounce,
...
-				prop: 'mapdata',
-				mpdgroups: groupsToLoad.join( '|' )
-			} ).then( function ( data ) {
-				var rawMapData = data.query.pages[ 0 ].mapdata,
-					i;
+		loader.nextFetch = [];
 
-				rawMapData = rawMapData && JSON.parse( rawMapData ) || {};
+		// FIXME: we need to fix this horrid hack
+		// http://stackoverflow.com/questions/39970101/combine-multiple-debounce-promises-in-js
```
Comment (juliengirault) [inline src/DataLoader.js]: 👍

**Review by juliengirault — CHANGES_REQUESTED** [review]

**Comment by juliengirault** [comment]
Agree, we should override [validateIndentation](http://jscs.info/rule/validateIndentation) in `.jscsrc`.

File: src/DataLoader.js (hunk 1 above)
Comment (nyurik) [reply src/DataLoader.js]: thx, fixed

**Comment by juliengirault** [comment]
Thanks !

**Pull request merged**